### PR TITLE
WIP: Eager mode

### PR DESCRIPTION
**Pull request by malmaud**

### CODE REVIEW

File: src/eager.jl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        finalizer(this) do self
+            @tfcall(:TFE_DeleteContext, Cvoid, (Ptr{Cvoid},), self.ptr)
+        end
+        @tfcall(:TFE_DeleteContextOptions, Cvoid, (Ptr{Cvoid},), options)
+        return this
+    end
+end
+
+
+mutable struct TensorHandle
```
Comment (oxinabox) [inline src/eager.jl]: this doesn't need to be mutable does it?

Comment (malmaud) [reply src/eager.jl]: I think it does so that it can have a finalizer.

File: src/eager.jl [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+end
+
+function setindex!(op::EagerOp, tensor::RawTensor, attr_name)
+    status = Status()
+    @tfcall(:TFE_OpSetAttrTensor, Cvoid, (Ptr{Cvoid}, Cstring, Ptr{Cvoid}, Ptr{Cvoid}), op, attr_name, tensor, status)
+    check_status(status)
+end
+
+function setindex!(op::EagerOp, dtype::DataType, attr_name)
+    @tfcall(:TFE_OpSetAttrType, Cvoid, (Ptr{Cvoid}, Cstring, TF_DataType), op, attr_name, dtype|>jl_to_df_type)
```
Comment (oxinabox) [inline src/eager.jl]:
```suggestion
    tf_dtype = jl_to_df_type(dtype)
    @tfcall(:TFE_OpSetAttrType, Cvoid, (Ptr{Cvoid}, Cstring, TF_DataType), op, attr_name, tf_dtype)
```

File: src/eager.jl [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    value = String(value)
+    @tfcall(:TFE_OpSetAttrString, Cvoid, (Ptr{Cvoid}, Cstring, Ptr{Cvoid}, Cint), op, attr_name, Vector{UInt8}(value), sizeof(value))
+end
+
+function setindex!(op::EagerOp, value::Vector, attr_name)
+    set_attr_list(op, attr_name, value)
+end
+
+function set_attr_list(op::EagerOp, attr_name, list::Vector{<:Integer})
+    list = Int64[Int64(x) for x in list]
```
Comment (oxinabox) [inline src/eager.jl]:
```suggestion
    list = Int64.(list)
```

File: src/eager.jl [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    @tfcall(:TFE_OpSetAttrIntList, Cvoid, (Ptr{Cvoid}, Cstring, Ptr{Int64}, Cint), op, attr_name, list, length(list))
+end
+
+function set_attr_list(op::EagerOp, attr_name, list::Vector{<:AbstractFloat})
+    list = Float32[Float32(x) for x in list]
```
Comment (oxinabox) [inline src/eager.jl]:
```suggestion
    list = Float32.(x)
```

File: src/eager.jl [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+function set_attr_list(op::EagerOp, attr_name, list::Vector{<:DataType})
+    list = map(jl_to_df_type, list)
+    @tfcall(:TFE_OpSetAttrTypeList, Cvoid, (Ptr{Cvoid}, Cstring, Ptr{Cvoid}, Cint), op, attr_name, list, length(list))
+end
+
+function set_attr_shape_list(op::EagerOp, attr_name, list::Vector)
+    dims = Vector{Int64}[]
+    for shape in list
+        push!(dims, Int64[shape...])
```
Comment (oxinabox) [inline src/eager.jl]: Isn't thiis an error?

File: src/eager.jl [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+end
+
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, c::EagerContext) = c.ptr
+
+function DeviceList(ctx::EagerContext)
+    status = Status()
+    ptr = @tfcall(:TFE_ContextListDevices, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}), ctx, status)
+    check_status(status)
+    count = @tfcall(:TF_DeviceListCount, Cint, (Ptr{Cvoid},), ptr)
+    this = new(ptr, count)
```
Comment (oxinabox) [inline src/eager.jl]: Can't call `new` outside of a inner constructor?

**Comment by malmaud** [comment]
Ya all that setattr stuff was jsut copy-pasted from core.jl and blindly tweaked. I suspect a lot of code on this branch will be ugly/buggy as I just try to get a basic system working, at which point I'll worry more about code quality.

**Comment by oxinabox** [comment]
OK, I'll stop bugging about it

File: src/crc.jl [hunk 7]
```diff
@@ -0,0 +1,87 @@
+module CRC
```
Comment (oxinabox) [inline src/crc.jl]: wat?
Why?

Also we have CRC in the stdlibs
https://github.com/JuliaLang/julia/blob/v1.1.0/stdlib/CRC32c/src/CRC32c.jl

**Comment by oxinabox** [comment]
Do you want to cherrypick this into like 4 PRs?

 - Eager Mode
 - Keras
 - TensorBoard
 - Everything else.

File: src/crc.jl (hunk 7 above)
Comment (malmaud) [reply src/crc.jl]: Oh I didn't know. Does it give exactly the same output as the implementation I have?

**Comment by malmaud** [comment]
Mmm if there's a real benefit. Right now I just want to get all the TensorFlow 2.0 features working and there is functionality shared between some of those parts so I'm not sure it's worth the effort given that all the code is going to be merged eventually anyways.

File: src/crc.jl (hunk 7 above)
Comment (oxinabox) [reply src/crc.jl]: Not sure,test it?

**Comment by oxinabox** [comment]
Ok, tell me when you are ready for review :-D

File: src/keras.jl [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+function fit(m::Sequential, x, y; n_epochs=1, batch_size=nothing)
+    optimizer = m.attrs["optimizer"]
+    for epoch in 1:n_epochs
+        tape = set_tape()
+        y_predicted = x
+        for layer in m.attrs["layers"]
+            y_predicted = forward(layer, y_predicted)
+        end
+        loss = m.attrs["loss"](y, y_predicted)
+        println("Epoch $epoch: Loss if $(item(loss))")
```
Comment (oxinabox) [inline src/keras.jl]:
```suggestion
        @info "" epoch loss=item(loss)
```

File: src/meta.jl [hunk 9]
```diff
@@ -170,3 +170,9 @@ macro tf(ex)
         end
     end |> esc
 end
+
+macro scalar_summary(f)
+    @capture(f, funcname(args__; kwargs__))
+
```
Comment (oxinabox) [inline src/meta.jl]:
```suggestion
```

File: src/meta.jl [hunk 10]
```diff
@@ -170,3 +170,9 @@ macro tf(ex)
         end
     end |> esc
 end
+
+macro scalar_summary(f)
+    @capture(f, funcname(args__; kwargs__))
+    
+end
+
```
Comment (oxinabox) [inline src/meta.jl]:
```suggestion
```

File: src/meta.jl [hunk 11]
```diff
@@ -170,3 +170,9 @@ macro tf(ex)
         end
     end |> esc
 end
+
+macro scalar_summary(f)
```
Comment (oxinabox) [inline src/meta.jl]: This is never used

File: src/ops.jl [hunk 12]
```diff
@@ -63,10 +67,10 @@ end
 macro define_broadcast(jl_op, tf_func)
```
Comment (oxinabox) [inline src/ops.jl]: One day, and that day is not today,
we need to go and fix how we do broadcasting.
And use broadcasting styles fully.

File: src/ops/sequences.jl [hunk 13]
```diff
@@ -15,7 +15,7 @@ end
 
 convert_eltype(x, dtype) = x
 
-@op function constant(value; dtype=nothing, kwargs...)
+@op function constant(value; dtype = nothing, kwargs...)
```
Comment (oxinabox) [inline src/ops/sequences.jl]: I wonder if we shouldn't have this return its own `Constant` type,
and have that type just type-promote into either a `Tensor` or a `TensorHandle` depending on what it touchs?

File: src/ops/summaries.jl [hunk 14]
```diff
@@ -49,4 +55,10 @@ function merge_all(key=:Summaries)
     merge(tensors)
 end
 
+macro scalar(f, args...)
```
Comment (oxinabox) [inline src/ops/summaries.jl]: This needs tests?

File: src/ops/transformations.jl [hunk 15]
```diff
@@ -381,13 +381,14 @@ Returns:
     with_op_name(name, "Transpose") do
         if perm === nothing
             r = range(constant(0), LinearAlgebra.rank(n)-1)
-            perm = reverse(r, [true])
+            perm = reverse(r, [0])
```
Comment (oxinabox) [inline src/ops/transformations.jl]: We should update `reverse(::Tensor, dims)` 
to match Julia's `reverse(::Tensor; dims)`

File: src/show.jl [hunk 16]
```diff
@@ -40,12 +40,19 @@ function Base.show(io::IO, t::RawTensor)
     end
 end
 
+function Base.show(io::IO, t::TensorHandle)
+    raw_tensor = resolve(t)
+    jl_array = convert(Array, raw_tensor)
```
Comment (oxinabox) [inline src/show.jl]: can just do `jl_array = convert(Array, t)`,

File: src/show.jl [hunk 17]
```diff
@@ -40,12 +40,19 @@ function Base.show(io::IO, t::RawTensor)
...
+    ptr = pointer_from_objref(t)
+    print(io, "Tensor<$ptr>($(jl_array))")
```
Comment (oxinabox) [inline src/show.jl]:
```suggestion
    print(io, "EagerTensor<$ptr>($(jl_array))")
```

or 

```suggestion
    print(io, "TensorHandle<$ptr>($(jl_array))")
```

File: src/summary_writer.jl [hunk 18]
```diff
@@ -28,28 +28,45 @@ Arguments:
 * logdir: A string. Directory where event file will be written.
 * graph: A `Graph` object.
 """
-function FileWriter(log_dir::AbstractString; graph=get_def_graph())
+function FileWriter(log_dir::AbstractString; graph=nothing)
+    if !tf.in_eager_mode() && graph===nothing
```
Comment (oxinabox) [inline src/summary_writer.jl]:
```suggestion
    if !tf.in_eager_mode() && graph === nothing
```

File: src/summary_writer.jl [hunk 19]
```diff
@@ -28,28 +28,45 @@ Arguments:
...
     mkpath(log_dir)
-    path = joinpath(log_dir, "events")
-    pyo = @py_proc pywrap_tensorflow[][:EventsWriter](py_bytes($path))
-    writer = FileWriter(pyo, String(log_dir))
+    local path
+    for i in Iterators.countfrom(1)
+        path = joinpath(log_dir, "events.out.tfevents.$i")
+        isfile(path) || break
+    end
+    rm(path, force=true)
```
Comment (oxinabox) [inline src/summary_writer.jl]:
```suggestion
```
Do not want this,
we've already proven there is no file of that name,
and if its a dir this won't remove it anyway since not recursive.

File: src/summary_writer.jl [hunk 20]
```diff
@@ -28,28 +28,45 @@ Arguments:
...
+        graph = get_def_graph()
+    end
     mkpath(log_dir)
-    path = joinpath(log_dir, "events")
-    pyo = @py_proc pywrap_tensorflow[][:EventsWriter](py_bytes($path))
-    writer = FileWriter(pyo, String(log_dir))
+    local path
+    for i in Iterators.countfrom(1)
+        path = joinpath(log_dir, "events.out.tfevents.$i")
+        isfile(path) || break
```
Comment (oxinabox) [inline src/summary_writer.jl]:
```suggestion
        ispath(path) || break
```

File: src/summary_writer.jl [hunk 21]
```diff
@@ -58,6 +75,12 @@ function Base.write(writer::FileWriter, summary::tensorflow.Summary, global_step
     setproperty!(event, :step, Int(global_step))
     setproperty!(event, :wall_time, time())
     setproperty!(event, :summary, summary)
+    # Some bug in ProtoBuf.jl is causing these to not be marked as filled,
```
Comment (oxinabox) [inline src/summary_writer.jl]: cross-link issue number?

File: src/summary_writer.jl [hunk 22]
```diff
@@ -76,7 +99,30 @@ function Base.write(writer::FileWriter, graph::Graph)
     write(writer, event)
 end
 
+function set_default(writer::FileWriter)
+    context = tf.Context()
+    context.attrs["default_file_writer"] = writer
+    push!(tf.global_context, context)
+end
+
+function with_default(writer::FileWriter, block)
```
Comment (oxinabox) [inline src/summary_writer.jl]:
```suggestion
function with_default(block, writer::FileWriter)
```

so that do-blocks work

File: src/summary_writer.jl [hunk 23]
```diff
@@ -76,7 +99,30 @@ function Base.write(writer::FileWriter, graph::Graph)
...
+    context = tf.Context()
+    context.attrs["default_file_writer"] = writer
+    tf.with_context(context, block)
```
Comment (oxinabox) [inline src/summary_writer.jl]:
```suggestion
    tf.with_context(block, context)
```

to match above change for do-blocks

File: src/tape.jl [hunk 24]
```diff
@@ -0,0 +1,171 @@
+using MacroTools
+import MacroTools: splitdef, combinedef
+
+mutable struct TapeNode
+    op::Function
+    args::Vector
+    results::Vector
+    kwargs::Dict
+end
```
Comment (oxinabox) [inline src/tape.jl]: Do we actually mutate this ever?

File: src/tape.jl [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+    if new_tape === nothing
+        new_tape = Tape()
+    end
+    context = Context()
+    context.attrs["tape"] = new_tape
+    push!(global_context, context)
+    return new_tape
+end
+
+function get_tape()
```
Comment (oxinabox) [inline src/tape.jl]: can and thus  should  be a 1 line function

File: src/tape.jl [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+    return global_context["tape"]
+end
+
+function add_node(t, node)
+    tape = get_tape()
+    tape === nothing && return
+    tape.nodes[t] = node
+end
+
+grad_fns = Dict()
```
Comment (oxinabox) [inline src/tape.jl]:
```suggestion
const  grad_fns = Dict()
```

if  we need to replace it can use `empty!`

Comment (oxinabox) [reply src/tape.jl]: But really why are we using a dict for this?
We can used dispatch for it.

Dispatching on the type of the function?

File: src/tape.jl [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+
+macro back_for(target, fn)
```
Comment (oxinabox) [inline src/tape.jl]: This should be be changed to instread make function methods
```
backwards(::typeof(Ops.add))= function(grad, x, y; kwargs...)
    println("Add got $grad, $x, $y")
    return [constant(1.0).*grad, constant(1.0).*grad]
end)
```
etc.

File: src/tape.jl [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+end)
+
+@back_for(Ops.pow, function f(grad, x, y; kwargs...)
+    [y.* (x.^(y.-1)), nothing]
+end)
+
+function with_no_grad(f)
+    context = Context()
+    context.attrs["tape"] = nothing
+    res = with_context(context, f)
```
Comment (oxinabox) [inline src/tape.jl]:
```suggestion
    res = with_context(f, context)
```

File: src/tape.jl [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+
+@back_for(Ops.neg, function f(grad, x; kwargs...)
+    return constant(-1.0) .* grad
+end)
+
+@back_for(Ops.pow, function f(grad, x, y; kwargs...)
+    [y.* (x.^(y.-1)), nothing]
+end)
+
+function with_no_grad(f)
```
Comment (oxinabox) [inline src/tape.jl]: Why is these here in the middle of all the gradient definitions?

File: src/summary_writer.jl [hunk 30]
```diff
@@ -4,7 +4,7 @@ import ..TensorFlow
 const tf = TensorFlow
 import ..TensorFlow: tensorflow, Graph, get_def_graph, @py_proc
 
-struct FileWriter
+struct FileWriter <: tf.Context
```
Comment (oxinabox) [inline src/summary_writer.jl]: :mindblown:

File: src/core.jl [hunk 31]
```diff
@@ -1227,6 +1233,8 @@ function Tensor(op::Operation, value_index::Int)
     Tensor{get_output_type(base_tensor)}(op, value_index)
 end
 
+# Tensor constructors
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/core.jl]: I'll just move them to the declaration of `struct Tensor` and the comment won't be necessary since that's where you'd expect constructors to be.

File: src/eager.jl [hunk 32]
```diff
@@ -0,0 +1,388 @@
+mutable struct EagerContext
+    ptr::Ptr{Cvoid}
+end
+
+@enum PlacementPolicy begin
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: done

File: src/eager.jl [hunk 33]
```diff
@@ -0,0 +1,388 @@
...
+    PLACEMENT_EXPLICIT=0
+    PLACEMENT_WARN=1
+    PLACEMENT_SILENT=2
+    PLACEMENT_SILENT_FOR_INT32=3
+end
+
+function EagerContext(;async=false, placement_policy=nothing)
+    # For some reason, this has to be called before :TFE_Execute or else tf
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: yes

File: src/eager.jl [hunk 34]
```diff
@@ -0,0 +1,388 @@
...
+    check_status(status)
+    count = @tfcall(:TF_DeviceListCount, Cint, (Ptr{Cvoid},), ptr)
+    this = DeviceList(ptr, count)
+    return this
+end
+
+mutable struct TensorHandle <: AbstractTensor{Any}
+    ptr::Ptr{Cvoid}
+
+    function TensorHandle(tensor::RawTensor)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: done

File: src/eager.jl [hunk 35]
```diff
@@ -0,0 +1,388 @@
...
+
+function resolve(h::TensorHandle)
+    status = Status()
+    ptr = @tfcall(:TFE_TensorHandleResolve, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}), h, status)
+    check_status(status)
+    tensor = RawTensor(ptr)
+    return tensor
+end
+
+function Base.Array(h::TensorHandle)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: I think this came up in trying to get tf getting to work with diffeq. but ya, lets remove all the hacks that were necessary for that and get that working in a maintainable way later.

File: src/eager.jl (hunk 5 above)
Comment (malmaud) [reply src/eager.jl]: I think it's right? At least Keno's PR https://github.com/malmaud/TensorFlow.jl/commit/932547509a9af0f73d7a1831078da0df7bc39d09 touched this code to make it what it is now.

File: src/eager.jl [hunk 36]
```diff
@@ -0,0 +1,388 @@
...
+
+function get_device(op::EagerOp)
+    status = Status()
+    str = @tfcall(:TFE_OpGetDevice, Cstring, (Ptr{Cvoid}, Ptr{Cvoid}), op, status)
+    return String(str)
+end
+
+function set_xla_compilation(op::EagerOp, enable)
+    @tfcall(:TFE_OpSetXLACompilation, Ptr{Cvoid}, (Ptr{Cvoid}, Cuchar), op, enable)
+    return enable
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: I'll add a docstring to clarify. A lot of doc strings can be added but that can come in the next PR.

File: src/eager.jl [hunk 37]
```diff
@@ -0,0 +1,388 @@
...
+end
+
+Base.convert(::Type{TensorHandle}, h::TensorHandle) = h
+Base.convert(::Type{TensorHandle}, h) = constant(h)
+
+function item(t::TensorHandle)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: 'item' here comes from the Pandas/PyTorch/xarray convention. But I'm open to debate.

File: src/eager.jl [hunk 38]
```diff
@@ -0,0 +1,388 @@
...
+    x = Array(t)
+    if length(x) != 1
+        throw(ErrorException("item can only be called on scalar tensors"))
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: I'll change the error name.

File: src/eager.jl [hunk 39]
```diff
@@ -0,0 +1,388 @@
...
+    return x[1]
+end
+
+Base.length(t::TensorHandle) = item(Ops.size(t))
+
+Base.eltype(::Type{TensorHandle}) = Float64 # temp hack
+Base.collect(t::TensorHandle) = Array(t)
+Base.iterate(t::TensorHandle, args...) = iterate(Array(t), args...)
+Base.zero(t::AbstractTensor) = Ops.zeros_like(t)
+Base.ones(t::AbstractTensor) = Ops.ones_like(t)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: We do. As you say, let's save that for a followup PR that removes it for both tensors in graph and eager mode.

File: src/eager.jl [hunk 40]
```diff
@@ -0,0 +1,388 @@
...
+
+function Base.:*(t1::TensorHandle, t2::Number)
+    return t1 .* t2
+end
+
+function Base.:*(t1::Number, t2::TensorHandle)
+    return t1 .* t2
+end
+
+function inplace_sub(x, y)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: 👍

File: src/eager.jl [hunk 41]
```diff
@@ -0,0 +1,388 @@
...
+function with_context(ctx, block)
+    push!(global_context, ctx)
+    res = block()
+    pop!(global_context)
+    return res
+end
+
+function get_eager_context()
+    return context_value("eager_context")
+end
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: Smart, I'll make that change.

File: src/generate_ops.jl [hunk 42]
```diff
@@ -114,14 +117,14 @@ function to_function(op::tensorflow.OpDef)
             end
         end
         if input._type > 0 && haskey(proto_type_map, input._type)
-            convert_target = tf.Tensor{proto_type_map[input._type]}
+            convert_target = (tf.Tensor{(proto_type_map[input._type])})
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: fixed

File: src/generate_ops.jl [hunk 43]
```diff
@@ -137,6 +140,8 @@ function to_function(op::tensorflow.OpDef)
             tf.add_input(desc, $(inputs[input_idx]))
         end)
     end
+    eager_input_block = input_block
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: At this point, the input block is the same so I can just delete that line.

File: src/generate_ops.jl [hunk 44]
```diff
@@ -174,6 +179,18 @@ function to_function(op::tensorflow.OpDef)
             end
         end)
     end
+    t_block = []
+    for (i, input_arg) in enumerate(op.input_arg)
+        if has_field(input_arg, :type_attr)
+            type_attr = input_arg.type_attr
+            if length(type_attr) > 0
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: I believe I added this guard because I had also assumed that and then it turned out it didn't.

File: src/ops/summaries.jl (hunk 14 above)
Comment (malmaud) [reply src/ops/summaries.jl]: added to summary_eager.jl

File: src/keras.jl [hunk 45]
```diff
@@ -0,0 +1,124 @@
...
+
+function forward(r::Relu, x)
+    nn.relu(x)
+end
+
+struct SGD
+    lr::TensorHandle
+end
+
+SGD(;lr=1e-3)= SGD(convert(TensorHandle, lr))
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/keras.jl]: This is mirroring keras. Let's make on-boarding easy :)

File: src/keras.jl [hunk 46]
```diff
@@ -0,0 +1,124 @@
...
+
+function Dense(in_size::Integer, out_size::Integer)
+    layer = Dense(constant(randn(in_size, out_size)), constant(zeros(out_size)))
+    return layer
+end
+
+@callable struct Relu <: Layer
+end
+
+function forward(r::Relu, x)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/keras.jl]: 👍🏻

File: src/eager.jl [hunk 47]
```diff
@@ -0,0 +1,388 @@
...
+
+function copy_to_device(ctx::EagerContext, h::TensorHandle, device_name)
+    status = Status()
+    res = TensorHandle()
+    res.ptr = @tfcall(:TFE_TensorHandleCopyToDevice, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}, Cstring, Ptr{Cvoid}), h, ctx, device_name, status)
+    check_status(status)
+    return res
+end
+
+copy_to_device(h, device_name) = copy_to_device(get_eager_context(), h, device_name)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/eager.jl]: `get_eager_context` will create a context now.

File: src/TensorFlow.jl [hunk 48]
```diff
@@ -141,8 +141,23 @@ function deallocator(data, len, arg)
 
 end
 
+struct Context
+    attrs::Dict
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/TensorFlow.jl]: done, by eliminating the whole type

File: src/TensorFlow.jl [hunk 49]
```diff
@@ -141,8 +141,23 @@ function deallocator(data, len, arg)
...
+
+Context() = Context(Dict())
+
+struct ContextStack
+    contexts::Vector{Context}
+end
+
+ContextStack() = ContextStack(Context[])
+
+global_context = ContextStack()
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/TensorFlow.jl]: done

File: src/core.jl [hunk 50]
```diff
@@ -521,6 +521,10 @@ mutable struct DeviceList
         end
         this
     end
+
+    function DeviceList(ptr, count)
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/core.jl]: done

File: src/core.jl [hunk 51]
```diff
@@ -1168,7 +1174,7 @@ function load_proto(value::tensorflow.AttrValue)
         load_proto(value.list)
     elseif has_field(value, :_type)
         type_ = value._type
-        proto_type_map[type_]
+        get(proto_type_map, type_, Float32)  # wrong
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/core.jl]: done

File: src/tape.jl (hunk 25 above)
Comment (malmaud) [reply src/tape.jl]: not applicable anymore since it's no longer a single expression. although if there's a way to write it more compactly, then we should do it.

File: src/tape.jl (hunk 26 above)
Comment (malmaud) [reply src/tape.jl]: Now we dispatch on function type.

File: src/tape.jl (hunk 28 above)
Comment (malmaud) [reply src/tape.jl]: done

File: src/tape.jl (hunk 29 above)
Comment (malmaud) [reply src/tape.jl]: fixed

File: src/summary_writer.jl (hunk 30 above)
Comment (malmaud) [reply src/summary_writer.jl]: Ya, we'll see if this is a good idea. It holds up for now though.

File: src/tape.jl (hunk 27 above)
Comment (malmaud) [reply src/tape.jl]: done

File: src/ops/transformations.jl (hunk 15 above)
Comment (malmaud) [reply src/ops/transformations.jl]: done

File: src/summary_writer.jl (hunk 21 above)
Comment (malmaud) [reply src/summary_writer.jl]: haven't filed one yet. should do that.

File: src/summary_writer.jl (hunk 22 above)
Comment (malmaud) [reply src/summary_writer.jl]: done. somehow always forget that julia takes the block first.

File: src/generate_ops.jl [hunk 52]
```diff
@@ -198,8 +215,20 @@ function to_function(op::tensorflow.OpDef)
...
+    eager_output_block = if scalar_output
+        quote
+            return res[1]
+        end
+    else
+        quote
+            #tf.execute(desc)
+            return res
+        end
+    end
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: thx, done

File: src/generate_ops.jl [hunk 53]
```diff
@@ -210,6 +239,47 @@ function to_function(op::tensorflow.OpDef)
...
+                $eager_output_block
+            end
+        end
+
+    end
+    call_kw_params = Expr(:parameters)
+    for arg in inputs[1].args
+        push!(call_kw_params.args, Expr(:kw, arg.args[1], arg.args[1]))
+    end
+    call_args = Any[call_kw_params]
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: done

File: src/generate_ops.jl [hunk 54]
```diff
@@ -210,6 +239,47 @@ function to_function(op::tensorflow.OpDef)
...
+    for input in inputs[2:end]
+        push!(call_args, input)
+    end
+    dispatch_expr = quote
+        function $jl_name($(inputs...))
+            if tf.in_eager_mode()
+                $(eager_name)($(call_args...))
+            else
+                $(graph_name)($(call_args...))
+            end
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/generate_ops.jl]: But what if the input is a native julia type and you are relying on the automatic promotion of julia types to tensors when an op is called? that's how it works now. we could deprecate that in a subsequent pr if we want.

File: src/keras.jl [hunk 55]
```diff
@@ -0,0 +1,124 @@
...
+macro callable(f)
+    name = struct_name(f)
+    quote
+        $(esc(f))
+        (m::$name)(args...; kwargs...) = forward(m, args...; kwargs...)
+    end
+end
+
+@callable mutable struct Sequential <: Model
+    attrs::Dict
```
[the remark this answers is not in the document]
Comment (malmaud) [reply src/keras.jl]: done.

File: src/summary_writer.jl (hunk 20 above)
Comment (oxinabox) [reply src/summary_writer.jl]: As I understand it
What we are actually looking for is the first available name that we can use without overwriting anything.

The normal thing to be getting in the way is old log files.

But it is also not strictly impossible the user has put a directory, a symlink, a mount point or something else there.

**Comment by malmaud** [comment]
Oh ya, good point.

**Comment by malmaud** [comment]
OK, ready for another review. Not everything is ideal, but as long as it's good enough, I would like to merge soon.

**Comment by oxinabox** [comment]
I will try and review this week

File: src/ops/sequences.jl (hunk 13 above)
Comment (malmaud) [reply src/ops/sequences.jl]: That would make sense, esp if we eliminate having a global eager mode. I think we can merge this for now and then worry about that, though.

**Comment by malmaud** [comment]
I think I'll just merge for now, but I still welcome comments if you get a chance/have the motivation sometime.

**Comment by oxinabox** [comment]
Fair, sorry I didn't get back to this
